Auto hide feature

SyarifFakhri opened the pull request:
- Added main functionality for auto hiding docks
- Auto hide toggle button is added in the dock area titlebar
- Side tab bars are added to the left and right of each window
- Add config for enabling the auto hide button, left side bar area and the right side bar area
- Add logic for save state and restore state with the auto hide functionality

githubuser0xFFFF commented on examples/autohide/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.5)
```
There is no autohide.pro file in this folder for building with qmake

githubuser0xFFFF commented on examples/centralwidget/autohide.pro:
```diff
@@ -2,7 +2,7 @@ ADS_OUT_ROOT = $${OUT_PWD}/../..
 
```
This file belongs into the autohide folder example. A centralwidget.pro file is missing

githubuser0xFFFF commented on examples/autohide/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.5)
+project(ads_example_centralwidget VERSION ${VERSION_SHORT}) 
```
The autohide example is missing in the examples.pro file

githubuser0xFFFF commented on src/DockContainerWidget.cpp:
```diff
@@ -1334,6 +1498,57 @@ CDockAreaWidget* CDockContainerWidget::addDockWidget(DockWidgetArea area, CDockW
+	}
+	if (!COverlayDockContainer::areaExistsInConfig(area))
+	{
+		assert(false, "Requested area does not exist in config");
```
User Q_ASSERT_X here instead of assert - this line does not build on Windows GCC

githubuser0xFFFF commented on src/OverlayDockContainer.cpp:
```diff
@@ -0,0 +1,315 @@
+/*******************************************************************************
```
This file is missing in src.pro

githubuser0xFFFF commented on src/OverlayDockContainer.h:
```diff
@@ -0,0 +1,134 @@
+#ifndef OverlayDockContainerH
```
This file is missing in src.pro

githubuser0xFFFF commented on src/OverlayDockContainer.cpp:
```diff
@@ -0,0 +1,315 @@
+#include <QBoxLayout>
+#include <QPainter>
+#include <QSplitter>
+
```
Compiling this files causes error: **OverlayDockContainer.cpp:52:32: error: field 'DockManager' has incomplete type 'QPointer<ads::CDockManager>'**

Add QPointer include:

```cpp
#include <QPointer>
```

githubuser0xFFFF commented on src/SideTabBar.h:
```diff
@@ -0,0 +1,81 @@
+#ifndef SideTabBarH
```
This file is missing in src.pro

githubuser0xFFFF commented on src/SideTabBar.cpp:
```diff
@@ -0,0 +1,101 @@
+/*******************************************************************************
```
This file is missing in src.pro

githubuser0xFFFF commented on src/DockWidgetSideTab.h:
```diff
@@ -0,0 +1,96 @@
+#ifndef DockWidgetSideTabH
```
This file is missing in src.pro

githubuser0xFFFF commented on src/DockWidgetSideTab.cpp:
```diff
@@ -0,0 +1,151 @@
+/*******************************************************************************
```
This file is missing in src.pro

githubuser0xFFFF reviewed (changes requested):
I fixed all build issues that occured when building your pull request on my Windows machine with qmake and pushed everything to a new branch **auto_hide_feature**

githubuser0xFFFF commented:
Thank you for this big merge request with a great new feature. I fixed all build issues in your pull request that occured when building everything on Windows with MinGW and qmake.

When I run the demo application to test the feature, I can see an issue that I will document in the next days.

githubuser0xFFFF commented:
The current implementation has a bug, that causes a wrong painting on Windows. To reproduce this, execute the following steps:

- Start the demo application AdvancedDockingSystemDemo.exe
- click the pin in the group highlighted in the following picture
![grafik](https://user-images.githubusercontent.com/4239304/188838764-64a3bd4a-9847-431c-aefd-67869104c8c7.png)
- now click the **Active X 0** or **Label 4** side tab - the sidebar widget is not paintet properly (no title bar)
![grafik](https://user-images.githubusercontent.com/4239304/188839239-531ce32b-40ee-4bb3-9b92-2d09f05271e4.png)
- if you click the **Label 1** side tab, then the sidebar is paintet properly:
![grafik](https://user-images.githubusercontent.com/4239304/188839510-5893c3c4-a6c2-4254-b09e-2118fdcee5ae.png)

Because you implemented the feature, it is easier for you to find the problem.

SyarifFakhri commented:
Thanks for reviewing it so quickly! I will merge this with the auto hide branch you created and look into the issues you brought up. Please let me know if you find anything else :). I'm still testing this, so I will post more bug fixes as I go along, will let you know once it's fully ready for review!

githubuser0xFFFF commented:
@SyarifFakhri 

Hi, I just pulled the latest commits from your development branch. 

I still can have a bug if I test it in the demo application. This is how to see the problem:

- Start the demo application AdvancedDockingSystemDemo.exe on Windows
- click the pin in the group highlighted in the following picture
![grafik](https://user-images.githubusercontent.com/4239304/188838764-64a3bd4a-9847-431c-aefd-67869104c8c7.png)
- now click the **Active X 0** or **Label 4** side tab - the sidebar widget is not paintet properly (no title bar)
![grafik](https://user-images.githubusercontent.com/4239304/189326003-106651e8-c4cc-46a4-b5a6-92d62fdd17a9.png)
- if you click the **Label 1** side tab, then the sidebar is paintet properly:
![grafik](https://user-images.githubusercontent.com/4239304/188839510-5893c3c4-a6c2-4254-b09e-2118fdcee5ae.png)

Is it a bug or do I need to modify or add something?

SyarifFakhri commented:
Sorry, this isn't ready for review yet. I haven't fixed the bug you posted. I think I must've requested review by mistake when trying to set this to draft. My apologies!

SyarifFakhri commented:
More changes from the initial:
- Fixed active X issue.
- Added a config for enabling auto hide, it is no longer enabled by default. Uncomment the config flag in `MainWindow.cpp` if you'd like to enable it in the demo.
```C++
    // CDockManager::setConfigFlag(CDockManager::DefaultAutoHideConfig, true);
```
- Side tab bars are added to the bottom as well. Left, right and bottom can all be enabled/disabled via config.
- Add a config for icons only, and added icon support.
- Fix numerous bugs and crashes. 

I sincerely apologize for the size of the PR. Please let me know if there's anything I can do to help or if you find any issues. Thank you for taking the time to review and test it :).

githubuser0xFFFF commented:
@SyarifFakhri  Hi, thank you for your contribution. I just tested the demo application and I see two issues. Here is how to reproduce:

- Start the demo application AdvancedDockingSystemDemo.exe
- click the pin in the group highlighted in the following picture

![grafik](https://user-images.githubusercontent.com/4239304/191707983-b2db7570-ea29-4353-9367-6eb1c0bb9a60.png)

- click the Label 4 dock widget side tab - the overlay is shown, but is is somehow transparent - that means , I do not see the label but widgets below the overlay (first issue)

![grafik](https://user-images.githubusercontent.com/4239304/191708709-a68f7d0a-febd-4231-8186-2a9642d8e2aa.png)

- now click the Label 4 pin button to unpin the Label 4 widget - the application crashes (second issue)

![grafik](https://user-images.githubusercontent.com/4239304/191709216-1ffc9cfe-b464-4489-a793-4a07402c1fd7.png)

SyarifFakhri commented:
> @SyarifFakhri Hi, thank you for your contribution. I just tested the demo application and I see two issues. Here is how to reproduce:
> 
>     * Start the demo application AdvancedDockingSystemDemo.exe
> 
>     * click the pin in the group highlighted in the following picture
> 
> 
> ![grafik](https://user-images.githubusercontent.com/4239304/191707983-b2db7570-ea29-4353-9367-6eb1c0bb9a60.png)
> 
>     * click the Label 4 dock widget side tab - the overlay is shown, but is is somehow transparent - that means , I do not see the label but widgets below the overlay (first issue)
> 
> 
> ![grafik](https://user-images.githubusercontent.com/4239304/191708709-a68f7d0a-febd-4231-8186-2a9642d8e2aa.png)
> 
>     * now click the Label 4 pin button to unpin the Label 4 widget - the application crashes (second issue)
> 
> 
> ![grafik](https://user-images.githubusercontent.com/4239304/191709216-1ffc9cfe-b464-4489-a793-4a07402c1fd7.png)

I've fixed this issue, let me know if it's fixed on your end :)

githubuser0xFFFF commented:
@SyarifFakhri Thank you for the fix. It fixed the issue so I can continuue testing. I found 2 other issues;

**First Issue:**

Dragging a dock widget will not hide an open auto hide widget. Steps to reproduce:

- open the demo application
- click the auto hide button of the **Label 0** dock widget.
- click the **Label 0** sidebar tab to show it
- drag the **Label 2** dock widget out of its position - this will not hide the open **Label 0** widget
   
![grafik](https://user-images.githubusercontent.com/4239304/194229671-8d5c56ea-ec60-4226-806b-31992e13d0ef.png)

If I test this in Visual Studio, then any click outside the auto hide widget or the start of a drag operatin hides the auto hide widget and I think, this should be the same here.

**Second Isse**

Switching between two auto hide widgets does not work properly. Steps to reproduce:

- open the demo application
- click the auto hide button of the **Label 0** dock widget.
- click the auto hide button of the **Filesystem 0** dock widget
- Now keep switching between the two Auto Hide widgets by alternately pressing the two Side Bar tabs - this will not work properly

SyarifFakhri commented:
> click the auto hide button of the Label 0 dock widget.

Actually the auto hide docks use focus switching in order to close the dock (losing focus causes it to close). However label 0 dock has `DockWidgetFocusable` set to `false`, so it can't be focused or unfocused. That's why it can only be closed via clicking the side widget tab as you saw.

 I considered this as an optional config; if the user doesn't want the dock to be hidden on losing focus, they can just set `DockWidgetFocusable` to `false`. I also put it in the documentation so users would know that this feature requires the dock widget being focusable if they want the dock widget to close on losing focus. What do you think?

githubuser0xFFFF commented:
I think, if you depend on the `DockWidgetFocusable` option of the DockWidget, then you should hide the pin button (auto hide button) for these widgets and prevent it from auto hiding it to prevent strange effects like in my reported issue.

githubuser0xFFFF commented:
@SyarifFakhri Thank you for your changes. I merged all your changes into the [auto_hide_feature branch](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/tree/auto_hide_feature). 

Becausen I don't really like the dependence of the Auto Hide feature on the Focus feature, I implemented some changes, that makes the Auto Hide feature work properly without the Focus feature. You can test the changes in the auto_hide_feature branch. 
I also renamed the OverlayDockContainer into AutoHideDockContainer to have a consistent naming and to avoid confusio with the `CDockOverlay` class.

One bug that is still there, also in your pull request: If I double click the header of the Auto Hide Widget, then the dock widget will be undocked but the SideTabWidget still remains i the side bar.

githubuser0xFFFF reviewed (commented):
@SyarifFakhri Thank you for your changes. I merged all your changes into the [auto_hide_feature branch](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/tree/auto_hide_feature). 

Because I don't really like the dependence of the Auto Hide feature on the Focus feature, I implemented some changes, that makes the Auto Hide feature work properly without the Focus feature. You can test the changes in the auto_hide_feature branch. The main logic is in the `CAutoHideDockContainer::eventFilter()` fucntion.
I also renamed the OverlayDockContainer into AutoHideDockContainer to have a consistent naming and to avoid confusion with the `CDockOverlay` class.

One bug that is still there, also in your pull request: If I double click the header of the Auto Hide Widget, then the dock widget will be undocked but the SideTabWidget still remains in the side bar.

SyarifFakhri commented:
Thanks for your quick replies and edits! I'll merge this into [auto_hide_feature branch](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/tree/auto_hide_feature). and fix the bug you mentioned

SyarifFakhri commented:
> Because I don't really like the dependence of the Auto Hide feature on the Focus feature, I implemented some changes, that makes the Auto Hide feature work properly without the Focus feature. You can test the changes in the auto_hide_feature branch. The main logic is in the CAutoHideDockContainer::eventFilter() fucntion.

I agree, I think this is better. I also fixed a small bug where you cannot click on left and bottom pinned splitters.

> I also renamed the OverlayDockContainer into AutoHideDockContainer

My bad for not noticing the name clash here with `CDockOverlay`. I changed function names and comments from overlay -> autohide everywhere else as well.

> If I double click the header of the Auto Hide Widget, then the dock widget will be undocked but the SideTabWidget still remains in the side bar.

Fixed! Let me know if you find any more issues.

githubuser0xFFFF commented:
@SyarifFakhri Just a question, What was the reason for not providing a Top sidebar? Currently ony left, right and bottom sidebars are available:

```c++
/**
     * Dock widget side tab bar locations
     */
    enum SideTabBarArea
    {
        LeftTop,
        LeftBottom,
        RightTop,
        RightBottom,
	Bottom
    };
```

I also noticed, that the stylesheet contains references to SideTabBarAreas "Left" and "Right" that does not longer exist:

```css
ads--CDockWidgetSideTab[sideTabBarArea="Left"] {
	border-left: 3px solid grey;
	border-bottom: 1px solid white;
}

ads--CDockWidgetSideTab[sideTabBarArea="Right"] {
	border-right: 3px solid grey;
	border-bottom: 1px solid white;
}
```

SyarifFakhri commented:
> Just a question, What was the reason for not providing a Top sidebar? Currently ony left, right and bottom sidebars are available:

Ah, I just didn't want to deal with any strange interactions between the titlebar and the topbar if there was any. Plus the PR was getting very big already. I can look into adding them.

> I also noticed, that the stylesheet contains references to SideTabBarAreas "Left" and "Right" that does not longer exist:

I'll remove these!

githubuser0xFFFF commented on src/DockWidgetSideTab.h:
```diff
@@ -0,0 +1,168 @@
+        LeftBottom,
+        RightTop,
+        RightBottom,
+		Bottom
```
Would be great, if later a Top area could be added

SyarifFakhri replied:
![image](https://user-images.githubusercontent.com/4475295/195859170-a9e9d5ca-def5-47f4-83c8-650765ad6848.png)

Added a top bar area!

SyarifFakhri replied:
FYI: The calculation of which tab bar area it should go to is in: `CDockContainerWidget::getDockAreaPosition`, but the general idea is, if width > height, it'll go to the top or bottom based on distance. Otherwise, left or right based on distance.

githubuser0xFFFF commented on src/DockWidgetSideTab.h:
```diff
@@ -0,0 +1,168 @@
+ * pinned dock widgets
+ */
+class ADS_EXPORT CDockWidgetSideTab : public QFrame
+{
```
It would be great, if the side tab would be more similar to the CDockWidgetTab - that means, that it has an isActiveTab() function and activeTab property. That means, a side tab bar may have one single active tab or no active tab. This would allow a styling of the tab depending on the visible state of the auto hide widget. 

So a tab would be more like a checkable action. If it is pressed (active) the auto hide dock widget is visible, if not, it is invisible.

SyarifFakhri replied:
![image](https://user-images.githubusercontent.com/4475295/195859810-bf3ba8ff-c1a9-42e7-9c00-ff8e8bf9c0c7.png)

Added! Also updated the styling so it's based on isActiveTab. I suppose you wanted to add this to remove the dependency on focusHighlighting?

githubuser0xFFFF commented:
@SyarifFakhri Thank you for adding the top area and for changing the CDockWidgetTab. I just did some bigger changes in the [auto_hide_feature](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/tree/auto_hide_feature) branch.

1. I changed the logic in the [moveContentsToParent](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/96a5443a0251f68b8317dc4c9c573408a96bc0f3/src/AutoHideDockContainer.cpp#L331) function. If an auto hide dock widget is unpinned, then it is inserted at its current location. This feels more natural and is less confusing for the user - in my opinion
2. I changed the logic in [toggleAutoHideArea](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/96a5443a0251f68b8317dc4c9c573408a96bc0f3/src/DockAreaWidget.cpp#L1103). Now only the current dock widget is converted into an auto hide widget and not all tabs in a dock area. I think this behaviour is more realistic for real world applications and is less destructive to the current layout.
3.  I replaced the sid tab areas `LeftTop, LeftBotton, RightTop` and `RightBottom` with the two areas `Left` and `Right` to reduce complexity. I do not see the need for a top and bottom sidebar for the left and right side.
4. I improved the logic for calculation of the side tab area depending on the position of the dock area. (https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/96a5443a0251f68b8317dc4c9c573408a96bc0f3/src/DockContainerWidget.cpp#L1546)
5. I added an alternative SVG icon for the pin button and added stylesheet code to show a normal icon in the dock area and a pinned icon in the CAutoHideDockContainer

Please test the changes. I did not testes on Linux yet.

SyarifFakhri commented:
@githubuser0xFFFF Thanks for taking a look and making all these changes! I will merge your branch and test it, but what do you think if I make these changes: 

RE: 2. Perhaps this could be added as a config? Similar to how the X button closing all dock widgets in an area can also be changed in the config.

RE: 3: I added this because it was needed for our application! It helps a lot in organization when adding dock widgets at runtime (which we do often - we add it at the bottom left and bottom right). Perhaps to reduce complexity, it doesn't have to be taken into account with the function `calculateSideTabBarArea`, but instead only being able to be added programmatically via the dock manager. 

RE: 4. I took a look at this function, I think it's better now and it makes a lot more sense! However, I noticed it doesn't take into account the disabled/enabled areas in the config. Should it take this into account? (Currently I made it so that the top, left, bottom, and right side tab bar areas can be individually enabled and disabled).

githubuser0xFFFF commented:
@SyarifFakhri 

> RE: 2. Perhaps this could be added as a config? Similar to how the X button closing all dock widgets in an area can also be changed in the config.

Ok, I added a config flag `AutoHideButtonTogglesArea`.

> 
> RE: 3: I added this because it was needed for our application! It helps a lot in organization when adding dock widgets at runtime (which we do often - we add it at the bottom left and bottom right). Perhaps to reduce complexity, it doesn't have to be taken into account with the function `calculateSideTabBarArea`, but instead only being able to be added programmatically via the dock manager.

Ok, I understand this. But for the moment I would prefer to keep it simple and stay with the four areas. Maybe we can add this later after the release of the auto hide feature. At the moments it smells very application specific for me and not very generic.

> 
> RE: 4. I took a look at this function, I think it's better now and it makes a lot more sense! However, I noticed it doesn't take into account the disabled/enabled areas in the config. Should it take this into account? (Currently I made it so that the top, left, bottom, and right side tab bar areas can be individually enabled and disabled).

Do you really think that people would disable certain areas? I do not really see the benefit in first enabling the auto hide feature and then disabling certain areas. Maybe we should drop the `DockContainerHasLeftSideBar..` flags and simply enable all side bars.?

githubuser0xFFFF commented:
@SyarifFakhri Something that still bothers me is, that the height of the side tab in the top and bottom sidebar is bigger than the height of the tabs in the left and right sidebar. Do you have an idea what could be the reason?

SyarifFakhri commented:
I merged your changes into this branch, and made some minor changes.

> Ok, I added a config flag AutoHideButtonTogglesArea.

Great!

> Maybe we can add this later after the release of the auto hide feature. At the moments it smells very application specific for me and not very generic.

Sounds good! We will probably open a separate branch later just for this, until you're ready to merge it :)

> Maybe we should drop the DockContainerHasLeftSideBar.. flags and simply enable all side bars.?

Agreed, I removed this.

> Something that still bothers me is, that the height of the side tab in the top and bottom sidebar is bigger than the height of the tabs in the left and right sidebar. 

All the logic for the margins of the sidetab is controlled in the [DockWidgetSideTab](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/60e0201060e530ef7a7b0d9f8bbdc538bf4c4fed/src/DockWidgetSideTab.cpp#L94)

I've updated the margins a little bit:

Old:
![image](https://user-images.githubusercontent.com/4475295/196317928-fca7aa12-a0d2-4c7e-8568-4c96791724f1.png)

New:
![image](https://user-images.githubusercontent.com/4475295/196316820-4056fe58-d40e-4c43-be2f-be9c905235b2.png)

I also updated the CSS a little bit to match the new enum without right bottom and left bottom.

SyarifFakhri commented:
> I would like to find out, why the horizontal side tab needs so much vertical space. The vertical side tab is much smaller and also the dock widget tabs need less space.

Well this part was just trial and error really. I just changed the spacing until I thought it looked aesthetically pleasing for the top and bottom side tab. 
With even less top spacing: 
![image](https://user-images.githubusercontent.com/4475295/196457937-55c982f1-e746-4651-a93c-f583baa830d0.png)
Would you prefer something like that?

> Btw. what is the reason fro wrapping the icon label of the side tab into a SideTabIconLabel widget? The class is not documented so I do not understand the requirement. The dock widget tab uses the icon label directly.

Ah there's really two reasons for that. 1. I want to have separate margins between the icon and the label (having the same margins looked weird) and 2. Calling `setContentsMargins` on a widget is ineffective; [I think this is well known](https://stackoverflow.com/questions/10660072/qwidget-setcontentsmargins-appears-to-be-ineffective-why-so). So I just decided it's neater to encapsulate the layout and the icon into one class, so I wouldn't have to have an extra layout data member in the `DockWidgetTab` class to keep track of. Then `setContentsMargins` on that class just forwards it to the layout internally.

githubuser0xFFFF commented:
@SyarifFakhri I decided to reimplement `CDockWidgetSideTab` based on QPushButton  - this gives a much cleaner code and I hope this does not break too much in your application. The interface did not change and eliding is not supported yet. But I'm not sure if eliding is required here, because I do not think, that people will have that much tabs in a side tab bar.

SyarifFakhri commented:
@githubuser0xFFFF I've merged all your changes and pushed it in this branch. I also fixed a missing dependency in the CMakeLists.txt, fixed the icon only config and did some small cleanup :)

githubuser0xFFFF commented on src/AutoHideDockContainer.cpp:
```diff
@@ -0,0 +1,527 @@
+
+	const auto dockSize = static_cast<int>(static_cast<float>(INT_MAX) * SplitterProportion);
+	const auto remainingSize = INT_MAX - dockSize;
+    switch (d->SideTabBarArea)
```
This code causes the following warning when building:

Warning: enumeration value 'Top' not handled in switch [-Wswitch]

githubuser0xFFFF replied:
The stylesheet seems to be broken somehow. If I run the application I can see many warnings like this:

`Warning: QPushButton(0x219ddbf89e0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddbf89e0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddca3f40, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddca3f00, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddcb4bb0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddcb49f0, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddbf89e0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddca3f40, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddca3f00, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddcb4bb0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddcb49f0, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddcb4270, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddcb4270, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219db0256e0, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1de0, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1fe0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1160, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddcb4270, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219db0256e0, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1de0, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1fe0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1160, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddce5160, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddce5160, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce7760, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce77a0, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce7c20, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce7a60, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddcb4270, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219db0256e0, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1de0, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1fe0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce1160, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddbf89e0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddca3f40, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddca3f00, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddcb4bb0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddcb49f0, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddce5160, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce7760, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce77a0, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce7c20, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddce7a60, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddee6820, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddee6820, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf97790, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf94f90, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf95150, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf95190, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddee6820, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf97790, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf94f90, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf95150, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf95190, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddfe4910, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddee6820, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddfe4910, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0285c0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0285c0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029a00, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029600, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029800, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddfbb5c0, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0285c0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029a00, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029600, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029800, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddfbb5c0, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddee6820, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddfe4910, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf97790, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf94f90, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf95150, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddf95190, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddfcbe70, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddfcbe70, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de059350, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0594d0, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0595d0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de059e90, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219ddfcbe70, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de059350, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0594d0, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0595d0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de059e90, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0285c0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029a00, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029600, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029800, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddfbb5c0, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de062160, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de062160, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029300, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de028b00, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0284c0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddc90b10, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de062160, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de029300, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de028b00, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0284c0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219ddc90b10, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0a2e60, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0285c0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0a2e60, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
DockWidget Label 5
Warning: QPushButton(0x219de064c20, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de062160, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de064c20, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de063220, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de063220, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9c90, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9650, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9a10, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9a50, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de063220, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9c90, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9650, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9a10, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0d9a50, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0eacc0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0eacc0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0ea100, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0e9300, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0e90c0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0e9500, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0eacc0, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0ea100, name="tabsMenuButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0e9300, name="detachGroupButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0e90c0, name="dockAreaAutoHideButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: ads::CTitleBarButton(0x219de0e9500, name="dockAreaCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de18df40, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de0a2e60, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))
Warning: QPushButton(0x219de18df40, name="tabCloseButton")  does not have a property named  "iconsize" ((null):0, (null))`

SyarifFakhri replied:
> Warning: enumeration value 'Top' not handled in switch

Fixed, thanks!

> The stylesheet seems to be broken somehow

I didn't get these warnings but I suspect it's because the case was different before. See if the latest push fixes it for you :)

I also fixed one more bug where right click detach of an overlay dock would not cause it to cleanup the overlay properly.

Let me know if you find anything else!

githubuser0xFFFF replied:
@SyarifFakhri Thank you for the fixes. I reimplemented the `CAutoHideDockContainer` based on a QFrame. That means the `CAutoHideDockContainer` is now a QFrame instead of a QSplitter and uses a custom `CResizeHandle` widget for resizing. So masking is not required anymore. I did this, because with the current QSplitter based implementation non opaque resizing was not possible because of the masked splitter.

Saving and restoring state for the new `CAutoHideDockContainer` is not completely implemented yet. If you like, you can fix the save and restore functionality for the new `CAutoHideDockContainer`.

 I also added a special `visual_studio_light.css` file for testing a visual studio like theme.

You will find all my changes in the **auto_hide_feature** branch.

SyarifFakhri replied:
> I reimplemented the CAutoHideDockContainer based on a QFrame. 

Actually this was my initial implementation, but I changed it to `QSplitter` after you reported issues with Active X (I'm not entirely sure why changing to `QSplitter` fixed it though). Do you get any issues with Active X using your `QFrame` implementation?

> Saving and restoring state for the new CAutoHideDockContainer is not completely implemented yet. If you like, you can fix the save and restore functionality for the new CAutoHideDockContainer.

I will take a look and implement this :)

> I also added a special visual_studio_light.css file for testing a visual studio like theme.

Cool!

I will merge all your changes and fix anything I find, thanks!

githubuser0xFFFF replied:
I implemented the save and restore function for the `CAutoHideDockContainer`. I did some initial tests. It seems to work but I see crashes and gliches when switching perspectives, so there seems to be some problems in the code outside `CAutoHideDockContainer`

githubuser0xFFFF replied:
The following Settings file will help you to reproduce the crash when loading perspectives. Download the settings file, rename it to Settings.ini and then copy it to your ADS demo application.

[Settings.txt](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/files/9870183/Settings.txt)

In the software open the perspective **FloatingAutoHide6**, then open the perspective **Default** and the open the perspective **FloatingAutoHide6** again - crash!!.

githubuser0xFFFF replied:
@SyarifFakhri  I fixed the cause of the crash. The problem is, the deletion of the SideTabBar will delete all tabs in this SideBar. If a perspective is loaded, FloatingWidgets may be deleted wich in turn delets its SideBars wich in turn deletes the tabs in this SideBar. The SideBar is not the owner of the tabs should outlive the deletion of the SideBar.

SyarifFakhri replied:
Yep, the true owner is the dockwidget, it's merely borrowed by the side tab bar. Thanks for the fix!

SyarifFakhri replied:
@githubuser0xFFFF I've merged all your changes and made some small ones:
1. Fixed a crash if you tried to add an auto hide container without it having a previous dockArea (e.g. if you added it programmatically on software start)
2. Related to above: I added a `setSize` function so that a user could resize the auto hide container on first add
3. Fixed the CMakeLists.txt (added ResizeHandle.h/cpp)
4. I added a resize margin so the user can't resize the auto hide container all the way to the end and get stuck

Thanks for your time and contributions so far! Please let me know if there's anything else to look into :)

SyarifFakhri commented:
Added one more fix for a crash I found. Not sure if this is related to auto hide, but when loading a configuration, dock widgets are not removed from their old dock areas. This can cause the CurrentIndex of the area to be out of bounds since the dock widget does not exist there anymore. If you then try to remove the dock widget it calls setCurrentIndex and can cause a crash. Only happens when loading certain configurations.  

Fix is to remove the dockwidgets from their old dock areas when restoring state.

githubuser0xFFFF commented on src/DockContainerWidget.cpp:
```diff
@@ -1189,6 +1189,12 @@ bool DockContainerWidgetPrivate::restoreDockArea(CDockingStateReader& s,
 			continue;
 		}
 
+		const auto oldDockArea = DockWidget->dockAreaWidget();
```
Could you please provide a small test case and instructions how to generate a crash here. I use the library since years in our product and have never seen any crash here so I would like to investigate the reason for this crash.

SyarifFakhri replied:
On further investigation, it is related to the auto hide feature. When restoring the [auto hide container](https://github.com/duerr-ndt/Qt-Advanced-Docking-System/blob/d56281e8600d3a9c9648afa7deb6674ae4654169/src/DockContainerWidget.cpp#L1118) the function `CAutoHideDockContainer::addDockWidget` internally [removes the old dock area](https://github.com/duerr-ndt/Qt-Advanced-Docking-System/blob/d56281e8600d3a9c9648afa7deb6674ae4654169/src/AutoHideDockContainer.cpp#L291). This is what triggers the crash. I guess you would never see the crash if you don't remove dock widgets while restoring (which was not done previously).

In this case, I'll preserve the old behavior, and simply not remove the dock widget in the autohide container when restoring in order to keep the behavior the same :)

SyarifFakhri replied:
Unfortunately, I can't get it to crash except with our application, but I will post a test case if I manage to.

githubuser0xFFFF commented:
@SyarifFakhri Please merge the changes from the [auto_hide_feature](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/tree/auto_hide_feature) branch. I did a lot of refactoring, renaming and cleanup to prepare for a release. Please merge and test everything if it works like expected.

SyarifFakhri commented on src/DockContainerWidget.cpp:
```diff
@@ -1547,11 +1547,10 @@ CAutoHideDockContainer* CDockContainerWidget::createAndSetupAutoHideContainer(
     DockWidget->sideTabWidget()->show();
 
-	auto AutoHideContainer = new CAutoHideDockContainer(DockWidget, area, this);
+	const auto AutoHideContainer = new CAutoHideDockContainer(DockWidget, area, this);
```
Prevents reassignment of the pointer, unlikely but const correct!

SyarifFakhri commented:
I did find one crash after merging that I did not get before:

1. If you start the demo application and then press save state:
![image](https://user-images.githubusercontent.com/4475295/199640890-01a91022-d8b9-46ac-91fe-3fc9195cc08c.png)

2. Then add a bunch of auto hide widgets (doesn't really matter where):
![image](https://user-images.githubusercontent.com/4475295/199640989-2f251b88-cfbc-4dd6-b8af-dfa88f307fed.png)

3. Then when clicking "restore state", I get a crash.
![image](https://user-images.githubusercontent.com/4475295/199641328-c15a7394-0ba5-4eac-ba68-9cd40a32a518.png)

Not sure what is causing it because it seems to be crashing in the native windows event handler. Did you make any changes you think could cause this?

SyarifFakhri commented:
I did find a fix for the crash, although it's a little bit of a workaround. The solution is to schedule the deletion of the dockArea after the deletion of the autohide area. This allows the pending native windows event handling to take place.

githubuser0xFFFF commented:
@SyarifFakhri Have you tested really the latest changes in the auto_hide_feature branch? I pushed some more changes yestarday. I cannot reproduce the crash here.

SyarifFakhri commented:
@githubuser0xFFFF I've pulled and updated your changes to the latest and tested it out again, everything looks good!

Thanks again for all your help and all the changes you made 👍

githubuser0xFFFF commented:
@SyarifFakhri Thank you for your contribution. I'm closing the pull request now because development now takes place in the autohide_feature branch.

FelixBer commented:
Just a heads up, QMouseEvent::globalPos is deprecated in Qt6:

https://doc.qt.io/qt-6/qmouseevent-obsolete.html#globalPos

Used twice around https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/7dee00f89b5d82a32e441d1ea9cf48cece966ef7/src/AutoHideDockContainer.cpp#L511

Also, 
#include "DockWidgetTab.h"
#include "DockingStateReader.h"
appear to be unused in AutoHideDockContainer.cpp

githubuser0xFFFF commented:
@FelixBer Thank you. Fixed it.

SyarifFakhri commented:
@githubuser0xFFFF 
Just something I noticed while building this:
1.
https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/8bc2236f4f252110b33889f048efe4b3f95277f8/src/AutoHideTab.h#L43

This gives a warning that class was previously declared as a struct 

2.
https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/8bc2236f4f252110b33889f048efe4b3f95277f8/demo/CMakeLists.txt#L15

I believe the `ImageViewer.cpp` and `RenderWidget.cpp` are missing in this CMakeLists.txt :)

githubuser0xFFFF commented:
@SyarifFakhri Thank you. Fixed it. Please test.

I have a question. In qmake I can add an RC file on Windows with the `RC_FILE` statement (see demo.pro `RC_FILE += app.rc`). Do you know how to add this for a CMake file?

SyarifFakhri commented:
Cheers! It does build now :)

> In qmake I can add an RC file on Windows with the RC_FILE statement (see demo.pro RC_FILE += app.rc). Do you know how to add this for a CMake file?

I believe you just need to add it like so: 
![image](https://user-images.githubusercontent.com/4475295/202992371-2e2bb4d5-564e-425e-a840-3f1e69be738f.png)

Let me know if that works for you :)

SyarifFakhri commented:
@githubuser0xFFFF I did find one more issue while testing this. If you open an auto hide dock tab and open the context menu as shown:
![image](https://user-images.githubusercontent.com/4475295/203196076-9234d477-9ca5-4d9c-97c6-75adc2c3e314.png)

Then click anywhere inside the dock but outside the context menu, the dock will collapse.

I believe that's due to this code here:

https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/blob/5a185268c44404e720fb751b011b13f8d3bc2d8b/src/AutoHideDockContainer.cpp#L507

The context menu is considered as a different window, so it closes. I think one fix would be to check if the widget is a descendant of the auto hide widget, and don't collapse it if it is perhaps?

githubuser0xFFFF commented:
@SyarifFakhri Thank you - good catch. I changed the implementation now. Now the auto hide container is not closed, if the user clicks into another floating window. If will only get closed in the following cases:

- user clicks in the same dock container but outside of the auto hide widget
- user clicks the auto hide tab
- user starts dragging a floating widget that is not the container of this auto hide widget
- user starts dragging a docked widget

Please test, if this is ok for you.

githubuser0xFFFF commented:
@SyarifFakhri  I noticed that there are still some things, that needs to get fixed with hiding the auto hide widget

githubuser0xFFFF commented:
@SyarifFakhri Ok, I think I finally fixed the eventFilter function ion auto hide widget to work properly. Please test.

yuriyoung commented:
how to pin/unpin(auto hide) a group?

githubuser0xFFFF commented:
@yuriyoung Just read the manual and play around with the demo application and you will find it out in some minutes.
